Change RevisionNumber to string and implemented relative TextBox

temi54c1l8 opened the pull request:
…evice Info

nimrof reviewed (changes requested):
Hi,
Thanks for the pull request.

Adding extra elements to the xsd would make it fail its validation so that is not the way, but where should the extra values be?

Do any of you know @CANopenNode @trojanobelix?
The xsd has been removed from CiA for some reason 😑

temi54c1l8 commented on libEDSsharp/eds.cs:
```diff
@@ -679,7 +679,7 @@ public partial class DeviceInfo : InfoSection
         /// </summary>
         [EdsExport]
-        public UInt32 RevisionNumber;
+        public string RevisionNumber="";
```
RevisionNumber is like ProductNumber, I used string because it can be represented as both decimal and hexadecimal

temi54c1l8 replied:
The tool always uses the string type for this data, the RevisionNumber (must have the major equal to OD 1018 subindex 3) was always set by default and could not be changed from the graphical interface

nimrof replied:
> RevisionNumber is like ProductNumber, I used string because it can be represented as both decimal and hexadecimal

I see, but how can we make sure it only contains a valid number and not a string?

temi54c1l8 replied:
The tool does not check this in any object, you can try for example with the ProductID, DeviceType, etc...
I changed the RevisionNumber to a string to make it conform to the rest of the existing code.

To solve your question, you should create a universal function that does this check and attach it to all objects, but it requires a major change (but this was not my goal)

CANopenNode replied:
I prefer to stay uint32 in database. GUI should be able to represent it as both decimal and hexadecimal.
However, `ProductNumber` should also be `uint32`...

See also our new ideas for those issues in #115

CANopenNode commented:
> The xsd has been removed from CiA for some reason 😑

@nimrof, what do you mean by that?

nimrof commented:
the xml schemas
https://www.can-cia.org/xml/1.0
https://www.can-cia.org/xml/1.1

CANopenNode commented:
Interesting. I think, `http://www.canopen.org/xml/1.1` is just a namespace for xml schema, actual schema does not need to be there.
However, there are still technical documents and schemas are there. Also for new CAN FD standard, there is only a standard for XDD (CiA1311) and not for EDS.

CANopenNode commented:
`FileRevision` is part of EDS specification and is type of UNSIGNED8. xml specification only has `fileVersion`, both 1.0 and 1.1.

`libEDSsharp/CanOpenXSD_1_1.cs` is auto generated from official schema. We can not add non standard elements here.

nimrof commented:
> Interesting. I think, `http://www.canopen.org/xml/1.1` is just a namespace for xml schema, actual schema does not need to be there.

Not sure what the xml standards says, but in this case the files used to be there.
https://web.archive.org/web/20240207232553/https://www.can-cia.org/xml/1.0/

> However, there are still technical documents and schemas are there. Also for new CAN FD standard, there is only a standard for XDD (CiA1311) and not for EDS.

Yes, but it is currently for paying members only, so out of reach for me at the moment😶

CANopenNode commented:
I'm a member, if you need some information I can look into the standards.

CANopenNode commented:
`RevisionNumber` could be added into XDD into one of the existing properties under `DeviceIdentity` class somehow. `libEDSsharp/CanOpenXDD_1_1.cs` could be edited for that.

trojanobelix commented:
I do not have access to the CiA DS311. But in 306 (EDS) and 301 RevisionNumber and ProductNuber are unsigned32. So I assume in DS311 it would be  unsigned32, too.

temi54c1l8 commented:
I changed the RevisionNumber back to UInt32 as requested, I set the display to hexadecimal to better differentiate the major and minor revisin.
I also implemented the OrderingCode in string as indicated in the CiA 306 1.4.0 specification.

CANopenNode commented:
We can not add additional fields into XSD file, because it is not according to the standard. You can put new fields into the standard fields somehow.

![device_identity_XDD](https://github.com/CANopenNode/CANopenEditor/assets/13575344/a1cfff23-a7c8-4162-a032-005dc08906d6)

temi54c1l8 commented:
RevisionNumber and OrderCode are in the [DeviceInfo] section.
I am not an expert in the graphics part, I close this pull request
